# Lambda Function Operations

In [9]:
import pandas as pd
import numpy as np 

In [11]:
df = pd.read_csv("Top 100 most Streamed - Sheet1.csv")
df.head()

title         artist                  top genre  year  \
0         Blinding Lights     The Weeknd  canadian contemporary r&b  2020   
1        Watermelon Sugar   Harry Styles                        pop  2019   
2  Mood (feat. iann dior)       24kGoldn                   cali rap  2021   
3       Someone You Loved  Lewis Capaldi                        pop  2019   
4                 Perfect     Ed Sheeran                        pop  2017   

   beats.per.minute  energy  danceability  loudness.dB  liveness  valance  \
0               171      73            51           -6         9       33   
1                95      82            55           -4        34       56   
2                91      72            70           -4        32       73   
3               110      41            50           -6        11       45   
4                95      45            60           -6        11       17   

   length  acousticness  speechiness  popularity  
0     200             0            6          91  
1     174            12            5          88  
2     141            17            4          88  
3     182            75            3          86  
4     263            16            2          86

In [13]:
df.shape

(100, 14)

In [35]:
df.isnull().sum()

title               0
artist              0
top genre           0
year                0
beats.per.minute    0
energy              0
danceability        0
loudness.dB         0
liveness            0
valance             0
length              0
acousticness        0
speechiness         0
popularity          0
dtype: int64

In [37]:
df.columns

Index(['title', 'artist', 'top genre', 'year', 'beats.per.minute', 'energy',
       'danceability', 'loudness.dB', 'liveness', 'valance', 'length',
       'acousticness', 'speechiness', 'popularity'],
      dtype='object')

In [39]:
df.dtypes

title               object
artist              object
top genre           object
year                 int64
beats.per.minute     int64
energy               int64
danceability         int64
loudness.dB          int64
liveness             int64
valance              int64
length               int64
acousticness         int64
speechiness          int64
popularity           int64
dtype: object

# Using Lamda function with apply

In [41]:
# max of popularity and energy
max_cols = df[['popularity','energy']].apply(np.max)
print(max_cols)

popularity    91
energy        92
dtype: int64


#### Apply on single column

Syntax<br>
lambda parameteres : expressions

In [43]:
max_pop = df[['popularity']].apply(lambda x : x.max())
min_pop = df[['popularity']].apply(lambda x : x.min())
print(max_pop)
print(min_pop)

popularity    91
dtype: int64
popularity    53
dtype: int64


In [47]:
max_live = df[['liveness']].apply(lambda X : X.max())
min_energy = df[['energy']].apply(lambda X : X.min())
print(max_live)
print(min_energy)

liveness    79
dtype: int64
energy    11
dtype: int64


#### Q1) Find the differnace between the max and min value of energy and liveness separately using apply function utilising the concept of lambda.

In [50]:
r1 = df[['energy','liveness']].apply(lambda X : X.max)
r1

energy      <bound method Series.max of 0     73\n1     82...
liveness    <bound method Series.max of 0      9\n1     34...
dtype: object

#### Q2) Find the differnace between the max and min value of loudness.dB, danceability and length separately using apply function utilising the concept of lambda.

In [60]:
r2 = df[['loudness.dB','danceability','length']].apply(lambda X : X.max() - X.min())
print(type(r2))
r2

<class 'pandas.core.series.Series'>


loudness.dB      11
danceability     56
length          235
dtype: int64

# assign()

1) Assign new columns to a dataframe.<br>
2) Rturns a new object with all original columns in addition to new ones. Existing columns that are re-aasigned will be overwitten.

Syntax<br>
assign(parameters,returns)

In [63]:
df.head(2)

title        artist                  top genre  year  \
0   Blinding Lights    The Weeknd  canadian contemporary r&b  2020   
1  Watermelon Sugar  Harry Styles                        pop  2019   

   beats.per.minute  energy  danceability  loudness.dB  liveness  valance  \
0               171      73            51           -6         9       33   
1                95      82            55           -4        34       56   

   length  acousticness  speechiness  popularity  
0     200             0            6          91  
1     174            12            5          88

#### Q3) Create a new column Revised_bpm where the beats per minute values are 110% of the  actual value. Use assign function utilising the concept of lambda.

bpm = beats.per.minute

In [69]:
df1 = df.assign(Revised_bpm = lambda X: df['beats.per.minute'] + 0.10 * df['beats.per.minute'])
df1.head()

title         artist                  top genre  year  \
0         Blinding Lights     The Weeknd  canadian contemporary r&b  2020   
1        Watermelon Sugar   Harry Styles                        pop  2019   
2  Mood (feat. iann dior)       24kGoldn                   cali rap  2021   
3       Someone You Loved  Lewis Capaldi                        pop  2019   
4                 Perfect     Ed Sheeran                        pop  2017   

   beats.per.minute  energy  danceability  loudness.dB  liveness  valance  \
0               171      73            51           -6         9       33   
1                95      82            55           -4        34       56   
2                91      72            70           -4        32       73   
3               110      41            50           -6        11       45   
4                95      45            60           -6        11       17   

   length  acousticness  speechiness  popularity  Revised_bpm  
0     200             0            6          91        188.1  
1     174            12            5          88        104.5  
2     141            17            4          88        100.1  
3     182            75            3          86        121.0  
4     263            16            2          86        104.5

In [71]:
df1[['title','artist','beats.per.minute','Revised_bpm','popularity']].head(7)

title           artist  beats.per.minute  Revised_bpm  \
0         Blinding Lights       The Weeknd               171        188.1   
1        Watermelon Sugar     Harry Styles                95        104.5   
2  Mood (feat. iann dior)         24kGoldn                91        100.1   
3       Someone You Loved    Lewis Capaldi               110        121.0   
4                 Perfect       Ed Sheeran                95        104.5   
5                Believer  Imagine Dragons               125        137.5   
6    lovely (with Khalid)    Billie Eilish               115        126.5   

   popularity  
0          91  
1          88  
2          88  
3          86  
4          86  
5          86  
6          86

In [73]:
df1.columns

Index(['title', 'artist', 'top genre', 'year', 'beats.per.minute', 'energy',
       'danceability', 'loudness.dB', 'liveness', 'valance', 'length',
       'acousticness', 'speechiness', 'popularity', 'Revised_bpm'],
      dtype='object')

In [75]:
df2 = df.assign(Energy_Dance_Ratio = lambda X: df['energy']/df['danceability'])
df2.head()

title         artist                  top genre  year  \
0         Blinding Lights     The Weeknd  canadian contemporary r&b  2020   
1        Watermelon Sugar   Harry Styles                        pop  2019   
2  Mood (feat. iann dior)       24kGoldn                   cali rap  2021   
3       Someone You Loved  Lewis Capaldi                        pop  2019   
4                 Perfect     Ed Sheeran                        pop  2017   

   beats.per.minute  energy  danceability  loudness.dB  liveness  valance  \
0               171      73            51           -6         9       33   
1                95      82            55           -4        34       56   
2                91      72            70           -4        32       73   
3               110      41            50           -6        11       45   
4                95      45            60           -6        11       17   

   length  acousticness  speechiness  popularity  Energy_Dance_Ratio  
0     200             0            6          91            1.431373  
1     174            12            5          88            1.490909  
2     141            17            4          88            1.028571  
3     182            75            3          86            0.820000  
4     263            16            2          86            0.750000

In [79]:
df2[['title','artist','top genre','year','energy','Energy_Dance_Ratio']].head(7)

title           artist                  top genre  year  \
0         Blinding Lights       The Weeknd  canadian contemporary r&b  2020   
1        Watermelon Sugar     Harry Styles                        pop  2019   
2  Mood (feat. iann dior)         24kGoldn                   cali rap  2021   
3       Someone You Loved    Lewis Capaldi                        pop  2019   
4                 Perfect       Ed Sheeran                        pop  2017   
5                Believer  Imagine Dragons                modern rock  2017   
6    lovely (with Khalid)    Billie Eilish                 electropop  2018   

   energy  Energy_Dance_Ratio  
0      73            1.431373  
1      82            1.490909  
2      72            1.028571  
3      41            0.820000  
4      45            0.750000  
5      78            1.000000  
6      30            0.857143

#### Creating multiple columns using assign

In [84]:
df3 =  df.assign(Revised_bpm = lambda x : 1.1 * df['beats.per.minute'], 
                 Energy_Dance_Ratio = lambda x : df['energy']/df['danceability'])
df3.head()

title         artist                  top genre  year  \
0         Blinding Lights     The Weeknd  canadian contemporary r&b  2020   
1        Watermelon Sugar   Harry Styles                        pop  2019   
2  Mood (feat. iann dior)       24kGoldn                   cali rap  2021   
3       Someone You Loved  Lewis Capaldi                        pop  2019   
4                 Perfect     Ed Sheeran                        pop  2017   

   beats.per.minute  energy  danceability  loudness.dB  liveness  valance  \
0               171      73            51           -6         9       33   
1                95      82            55           -4        34       56   
2                91      72            70           -4        32       73   
3               110      41            50           -6        11       45   
4                95      45            60           -6        11       17   

   length  acousticness  speechiness  popularity  Revised_bpm  \
0     200             0            6          91        188.1   
1     174            12            5          88        104.5   
2     141            17            4          88        100.1   
3     182            75            3          86        121.0   
4     263            16            2          86        104.5   

   Energy_Dance_Ratio  
0            1.431373  
1            1.490909  
2            1.028571  
3            0.820000  
4            0.750000

In [87]:
df3.shape

(100, 16)

In [89]:
print(df.columns)

Index(['title', 'artist', 'top genre', 'year', 'beats.per.minute', 'energy',
       'danceability', 'loudness.dB', 'liveness', 'valance', 'length',
       'acousticness', 'speechiness', 'popularity'],
      dtype='object')


In [93]:
df3[['title','artist','top genre','year','energy','danceability','Energy_Dance_Ratio','beats.per.minute','Revised_bpm']].head(7)

title           artist                  top genre  year  \
0         Blinding Lights       The Weeknd  canadian contemporary r&b  2020   
1        Watermelon Sugar     Harry Styles                        pop  2019   
2  Mood (feat. iann dior)         24kGoldn                   cali rap  2021   
3       Someone You Loved    Lewis Capaldi                        pop  2019   
4                 Perfect       Ed Sheeran                        pop  2017   
5                Believer  Imagine Dragons                modern rock  2017   
6    lovely (with Khalid)    Billie Eilish                 electropop  2018   

   energy  danceability  Energy_Dance_Ratio  beats.per.minute  Revised_bpm  
0      73            51            1.431373               171        188.1  
1      82            55            1.490909                95        104.5  
2      72            70            1.028571                91        100.1  
3      41            50            0.820000               110        121.0  
4      45            60            0.750000                95        104.5  
5      78            78            1.000000               125        137.5  
6      30            35            0.857143               115        126.5